src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,484 @@
+  AdjointMsg BuildZerosAdjointNested(const StructInfo& sinfo) {
+    return MapToNestedMsg<Expr>(sinfo, [](StructInfo sinfo) {
+      auto tensor_sinfo = sinfo.as<TensorStructInfoNode>();
+      ICHECK(tensor_sinfo) << "The leaf of adjoint should be a Tensor.";
+      ICHECK(tensor_sinfo->shape.defined()) << "Error: missing shape when building zeros tuple.";
+      const Expr& init = zeros(tensor_sinfo->shape.value(), tensor_sinfo->dtype);
+      UpdateStructInfo(init, sinfo);
+      return init;
+    });
+  }
+
+  // Init the gradient of the var as ones op and update it in adjoint_msg_map_.
+  static AdjointMsg BuildOnesAdjoint(const Var& var) {
+    auto tensor_sinfo = GetStructInfoAs<TensorStructInfoNode>(var);
+    ICHECK(tensor_sinfo) << "The leaf of adjoint should be a Tensor.";
+    ICHECK(tensor_sinfo->shape.defined()) << "Error: missing shape when building ones.";
+    const Expr& init = ones(tensor_sinfo->shape.value(), tensor_sinfo->dtype);
+    UpdateStructInfo(init, GetRef<StructInfo>(tensor_sinfo));
+    return init;
+  }
+
+  // Return base + increment. A tuple-aware addition.
+  AdjointMsg TupleAwareAdd(const AdjointMsg& base, const AdjointMsg& increment) {
+    return CombineNestedMsg(base, increment, [&](Expr lhs, Expr rhs) {
+      if (IsCallZeros(lhs)) {
+        return rhs;
+      } else if (IsCallZeros(rhs)) {
+        return lhs;
+      }
+      auto sinfo = GetStructInfoAs<TensorStructInfoNode>(lhs);
+      ICHECK(sinfo) << "The leaf of adjoint should have StructInfo and be a Tensor.";
+      ICHECK(GetStructInfoAs<TensorStructInfoNode>(rhs))
+          << "The leaf of adjoint should have StructInfo and be a Tensor.";
+      Expr res = add(lhs, rhs);
+      UpdateStructInfo(res, GetRef<StructInfo>(sinfo));
+      return res;
+    });
+  }
+
+  // Perform an addition in a specified position of tuple.
+  // e.g. tuple=(a, b, c), index=1, increment=d, then return (a, b+d, c)
+  AdjointMsg AddElementInNestedAdjoint(const AdjointMsg& adjoint, int index,
+                                       const AdjointMsg& increment) {
+    ICHECK(adjoint.IsNested()) << "The adjoint should be nested.";
+    Array<AdjointMsg> arr = adjoint.NestedArray();
+    Array<AdjointMsg> new_arr;
+    new_arr.reserve(arr.size());
+    for (size_t i = 0; i < arr.size(); ++i) {
+      if (static_cast<int>(i) == index) {
+        new_arr.push_back(TupleAwareAdd(arr[i], increment));
+      } else {
+        new_arr.push_back(arr[i]);
+      }
+    }
+    return AdjointMsg(new_arr);
+  }
+
+  Expr AdjointMsgToExpr(AdjointMsg msg) {
+    auto fmapmerge = [](Array<Expr> subexpr) { return Tuple(subexpr); };
+    auto fmapleaf = [](AdjointMsg leaf_msg) {
+      ICHECK(leaf_msg.IsLeaf());
+      return leaf_msg.LeafValue();
+    };
+    auto fmapnull = []() {
+      LOG(FATAL) << "Error: null field should not appear in adjoint.";
+      return Expr();  // just for type checking
+    };
+    return MapFromNestedMsg<Expr>(msg, fmapmerge, fmapleaf, fmapnull);
```

Comment:
I think this all looks good to be default behavior and don't need to specifically extract it out as a configurable template.

Also when creating Tuple(subexpr), it might be great to check if Tuple(subexpr) can be simplified, e.g. Tuple(t[0], t[1]) where t only has 2 fields.

Reply:
Do you mean just handle the case `T=Expr`, and remove three map functions `fmapmerge`, `fmapleaf` and `fmapnull`? But I think this general version may have some other applications for different `NestedMsg<T>`. E.g. when we track some info like `NestedMsg<Int>`, we may finally want a Int value defined by this MapFromNestedMsg:
- fmapmerge: sum in this array
- fmapleaf: return the leaf value
- fmapnull: return 0
And there are also many other cases just like segment-tree algorithm.

Reply:
Another thing is that here we don't want the leaf msg to be `null` but in other scenarios we may want  to map `null` to some specified value.

Reply:
I think we could support two purposes
- Supporting `NestedMsg<T>` for general T. Like int, Array, etc
- Give the default behavior for `NestedMsg<Expr>`. That is transforming NestedMsg into Tuples

Maybe we could add functions to support both cases

Reply:
In our case, we only require the return Type to be Expr, and I think it's good now to just support this case.

Also, generally, I think we don't want to map null but eliminate null in msgs. In our case, we can assign 0 to vars that are not propagated in the adjoint map to avoid null, I think.

In the Expr case, fmerge is good to be default to be compose tuples simply.